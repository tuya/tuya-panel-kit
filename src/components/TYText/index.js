import React from 'react';
import PropTypes from 'prop-types';
import { Text, ColorPropType, StyleSheet } from 'react-native';
import StyledText from './styled';

/**
 * TODO:
 * 1. Text 三类 heading title paragraph，通过 props 去做还是通过不同组件去做;
 * 2. 是否还有更多属性需要暴露;
 */
class TYText extends React.Component {
  static propTypes = {
    /**
     * 字体类型
     */
    type: PropTypes.oneOf(['heading', 'title', 'paragraph']),
    /**
     * 字体尺寸
     */
    size: PropTypes.oneOfType([PropTypes.oneOf(['large', 'normal', 'small']), PropTypes.number]),
    /**
     * 字体对齐方式
     */
    align: PropTypes.oneOf(['left', 'center', 'right']),
    /**
     * 字体粗细
     */
    weight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    /**
     * 字体颜色
     */
    color: ColorPropType,
    ...Text.propTypes,
  };

  static defaultProps = {
    type: null,
    size: null,
    align: null,
    weight: null,
    color: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      text: props.text ? props.text : props.children,
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ text: nextProps.text ? nextProps.text : nextProps.children });
  }

  setText = text => {
    this.setState({ text });
  };

  setNativeProps = nativeProps => {
    this._text.setNativeProps(nativeProps);
  };

  render() {
    const { style, size, ...props } = this.props;
    /**
     * 如果 `style` 里有设置 `fontSize` 则覆盖 `size props`，
     * 避免行高 `lineHeight` 仍有误;
     */
    const { fontSize } = StyleSheet.flatten([style]);
    const realSize = fontSize || size;
    return (
      <StyledText
        style={style}
        ref={ref => {
          this._text = ref;
        }}
        allowFontScaling={false}
        size={realSize}
        {...props}
      >
        {this.state.text}
      </StyledText>
    );
  }
}
export default TYText;
