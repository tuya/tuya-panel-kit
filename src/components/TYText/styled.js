import { Text } from 'react-native';
import styled, { css } from 'styled-components/native';
import { defaultTheme } from '../theme';
import CoreUtils from '../../utils/core';

const { get } = CoreUtils;

const sizeStyles = props => {
  if (typeof props.size === 'number') {
    return css`
      font-size: ${props.size}px;
    `;
  }
  if (!props.type || !props.size) {
    return null;
  }
  const path = `text.${props.type}.${props.size}`;
  const ret = get(props.theme, path, defaultTheme.text.title.normal);
  if (typeof ret === 'function') {
    const result = ret(props);
    const cssStyle = css`
      font-size: ${result.fontSize}px;
      line-height: ${result.lineHeight};
    `;
    return cssStyle;
  }
  const cssStyle = css`
    font-size: ${ret.fontSize}px;
    line-height: ${ret.lineHeight};
  `;
  return cssStyle;
};

const colorStyle = css`
  color: ${props => {
    const type = get(props.theme, 'type', 'light');
    const path = `global.text.${type}`;
    return props.color || get(props.theme, path, defaultTheme.global.text.light);
  }};
`;

const textAlignStyle = css`
  text-align: ${props => props.align};
`;

const weightStyle = css`
  font-weight: ${props => props.weight};
`;

const fontFamilyStyle = css`
  font-family: ${props => {
    return get(props.theme, 'global.fontFamily', defaultTheme.global.fontFamily);
  }};
`;

export default styled(Text)`
  ${colorStyle};
  ${props => props.align && textAlignStyle};
  ${props => props.weight && weightStyle};
  ${props => sizeStyles(props)};
  background-color: transparent;
  ${props =>
    typeof get(props.theme, 'global.fontFamily', defaultTheme.global.fontFamily) === 'string' &&
    fontFamilyStyle}
`;
