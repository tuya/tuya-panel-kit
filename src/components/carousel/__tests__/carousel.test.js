import React from 'react';
import { View, Platform, ViewPagerAndroid, ScrollView } from 'react-native';
import { shallow } from 'enzyme';
import Carousel from '../index';
import TYText from '../../TYText';

const carouselChange = jest.fn();

function setup(props = {}) {
  const wrapper = shallow(
    <Carousel
      style={{ height: 180 }}
      selectedIndex={0}
      hasDots={true}
      loop={true}
      carouselChange={carouselChange}
      dotStyle={{ width: 8, height: 8, borderRadius: 4, backgroundColor: '#fff' }}
      dotActiveStyle={{ backgroundColor: '#ff0' }}
      {...props}
    >
      <View style={{ flex: 1, backgroundColor: 'red' }}>
        <TYText style={{ color: '#333' }}>Carousel 1</TYText>
      </View>
      <View style={{ flex: 1, backgroundColor: 'blue' }}>
        <TYText style={{ color: '#333' }}>Carousel 2</TYText>
      </View>
      <View style={{ flex: 1, backgroundColor: 'yellow' }}>
        <TYText style={{ color: '#333' }}>Carousel 3</TYText>
      </View>
    </Carousel>
  );
  const instance = wrapper.instance();
  return { wrapper, instance };
}

describe('Carousel Component', () => {
  jest.useFakeTimers();
  it('renders correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
    expect(wrapper).toMatchSnapshot();
  });
  it('autoplay methods with count(2) and os(ios)', () => {
    Platform.OS = 'ios';
    const { wrapper, instance } = setup({
      autoplay: true,
      loop: false,
      onMomentumScrollEnd: jest.fn(),
    });
    instance.count = 2;
    wrapper.setState({ isScrolling: true }, () => {
      jest.runOnlyPendingTimers();
      jest.runOnlyPendingTimers();
    });
  });
  it('autoplay methods with count(2) and os(android)', () => {
    Platform.OS = 'android';
    let wrapper = setup({
      autoplay: true,
      loop: false,
      onMomentumScrollEnd: jest.fn(),
      useViewPagerOnAndroid: false,
    });
    wrapper.instance.count = 2;
    wrapper.instance.scrollview = {
      setPageWithoutAnimation: jest.fn(),
      setPage: jest.fn(),
      scrollTo: jest.fn(),
    };
    wrapper.wrapper.setState({ loopJump: true }, () => {
      jest.runOnlyPendingTimers();
      jest.runOnlyPendingTimers();
    });
    wrapper = setup({
      autoplay: true,
      loop: false,
      onMomentumScrollEnd: jest.fn(),
      useViewPagerOnAndroid: true,
    });
    wrapper.instance.count = 2;
    wrapper.instance.scrollview = {
      setPageWithoutAnimation: jest.fn(),
      setPage: jest.fn(),
      scrollTo: jest.fn(),
    };
    wrapper.wrapper.setState({ loopJump: true }, () => {
      jest.runOnlyPendingTimers();
      jest.runOnlyPendingTimers();
    });
  });
  it('autoplay methods with count(1)', () => {
    const { instance } = setup({ autoplay: true, loop: false });
    instance.count = 1;
    jest.runOnlyPendingTimers();
  });
  it('componentWillReceiveProps with selectedIndex(100) ', () => {
    const { wrapper } = setup();
    wrapper.setProps({ selectedIndex: 100 });
  });
  it('componentWillReceiveProps with selectedIndex(0) ', () => {
    const { wrapper, instance } = setup();
    instance.count = 100;
    wrapper.setProps({ selectedIndex: 0 });
  });
  it('trigger onPageScrollStateChanged', () => {
    Platform.OS = 'android';
    const { wrapper } = setup({ onScrollBeginDrag: jest.fn() });
    const viewPagerAndroid = wrapper.find(ViewPagerAndroid);
    viewPagerAndroid.simulate('pageScrollStateChanged', 'dragging');
  });
  it('trigger onScrollEndDrag', () => {
    Platform.OS = 'ios';
    const { wrapper } = setup({ onScrollEndDrag: jest.fn() });
    const scrollView = wrapper.find(ScrollView);
    wrapper.setState({ selectedIndex: 0 }, () => {
      scrollView.simulate('scrollEndDrag', { nativeEvent: { contentOffset: { x: 0 } } });
    });
  });
  it('trigger onLayout', () => {
    Platform.OS = 'android';
    const { wrapper, instance } = setup({ useViewPagerOnAndroid: false });
    const targetNode = wrapper.findWhere(
      node => node.name() === 'View' && !!node.prop('onLayout') === true
    );
    instance.scrollview = { scrollTo: jest.fn() };
    targetNode.simulate('layout', { nativeEvent: { layout: { width: 100 } } });
    jest.runOnlyPendingTimers();
  });
  it('trigger onScrollEnd', () => {
    Platform.OS = 'android';
    const { wrapper } = setup({ useViewPagerOnAndroid: true });
    const targetNode = wrapper.find(ViewPagerAndroid);
    targetNode.simulate('pageSelected', {
      nativeEvent: {
        contentOffset: { x: -1 },
      },
    });
    wrapper.setState({ width: 1 }, () => {
      targetNode.simulate('pageSelected', {
        nativeEvent: {
          contentOffset: { x: 1 },
        },
      });
    });
  });
  it('renders correctly with one child', () => {
    const wrapper = shallow(
      <Carousel
        style={{ height: 180 }}
        selectedIndex={0}
        hasDots={true}
        loop={true}
        carouselChange={carouselChange}
        dotStyle={{ width: 8, height: 8, borderRadius: 4, backgroundColor: '#fff' }}
        dotActiveStyle={{ backgroundColor: '#ff0' }}
      >
        <View style={{ flex: 1, backgroundColor: 'red' }}>
          <TYText style={{ color: '#333' }}>Carousel 1</TYText>
        </View>
      </Carousel>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly with no child', () => {
    const wrapper = shallow(
      <Carousel
        style={{ height: 180 }}
        selectedIndex={0}
        hasDots={true}
        loop={true}
        carouselChange={carouselChange}
        dotStyle={{ width: 8, height: 8, borderRadius: 4, backgroundColor: '#fff' }}
        dotActiveStyle={{ backgroundColor: '#ff0' }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
