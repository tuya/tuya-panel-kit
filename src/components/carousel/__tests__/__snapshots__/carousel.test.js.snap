// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Component renders correctly 1`] = `
<View
  onLayout={[Function]}
>
  <ScrollViewMock
    accessibilityLabel="Carousel"
    automaticallyAdjustContentInsets={false}
    autoplay={false}
    autoplayInterval={2000}
    bounces={true}
    carouselChange={[MockFunction]}
    contentContainerStyle={
      Object {
        "height": 180,
      }
    }
    contentOffset={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    directionalLockEnabled={true}
    dotActiveStyle={
      Object {
        "backgroundColor": "#ff0",
      }
    }
    dotStyle={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 4,
        "height": 8,
        "width": 8,
      }
    }
    dots={[Function]}
    hasDots={true}
    horizontal={true}
    loop={true}
    onMomentumScrollEnd={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    pageStyle={Object {}}
    pagingEnabled={true}
    removeClippedSubviews={false}
    selectedIndex={0}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    style={
      Object {
        "height": 180,
      }
    }
    useViewPagerOnAndroid={true}
  >
    <View
      accessibilityLabel="Carousel_0"
      key="carousel_0"
      style={
        Array [
          Object {},
          Object {
            "width": 0,
          },
        ]
      }
    >
      <View
        key=".2"
        style={
          Object {
            "backgroundColor": "yellow",
            "flex": 1,
          }
        }
      >
        <TYText
          align={null}
          color={null}
          size={null}
          style={
            Object {
              "color": "#333",
            }
          }
          type={null}
          weight={null}
        >
          Carousel 3
        </TYText>
      </View>
    </View>
    <View
      accessibilityLabel="Carousel_1"
      key="carousel_1"
      style={
        Array [
          Object {},
          Object {
            "width": 0,
          },
        ]
      }
    >
      <View
        key=".0"
        style={
          Object {
            "backgroundColor": "red",
            "flex": 1,
          }
        }
      >
        <TYText
          align={null}
          color={null}
          size={null}
          style={
            Object {
              "color": "#333",
            }
          }
          type={null}
          weight={null}
        >
          Carousel 1
        </TYText>
      </View>
    </View>
    <View
      accessibilityLabel="Carousel_2"
      key="carousel_2"
      style={
        Array [
          Object {},
          Object {
            "width": 0,
          },
        ]
      }
    >
      <View
        key=".1"
        style={
          Object {
            "backgroundColor": "blue",
            "flex": 1,
          }
        }
      >
        <TYText
          align={null}
          color={null}
          size={null}
          style={
            Object {
              "color": "#333",
            }
          }
          type={null}
          weight={null}
        >
          Carousel 2
        </TYText>
      </View>
    </View>
    <View
      accessibilityLabel="Carousel_3"
      key="carousel_3"
      style={
        Array [
          Object {},
          Object {
            "width": 0,
          },
        ]
      }
    >
      <View
        key=".2"
        style={
          Object {
            "backgroundColor": "yellow",
            "flex": 1,
          }
        }
      >
        <TYText
          align={null}
          color={null}
          size={null}
          style={
            Object {
              "color": "#333",
            }
          }
          type={null}
          weight={null}
        >
          Carousel 3
        </TYText>
      </View>
    </View>
    <View
      accessibilityLabel="Carousel_4"
      key="carousel_4"
      style={
        Array [
          Object {},
          Object {
            "width": 0,
          },
        ]
      }
    >
      <View
        key=".0"
        style={
          Object {
            "backgroundColor": "red",
            "flex": 1,
          }
        }
      >
        <TYText
          align={null}
          color={null}
          size={null}
          style={
            Object {
              "color": "#333",
            }
          }
          type={null}
          weight={null}
        >
          Carousel 1
        </TYText>
      </View>
    </View>
  </ScrollViewMock>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "bottom": 10,
          "left": 0,
          "position": "absolute",
          "right": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        key="dot-0"
        style={
          Array [
            Object {
              "backgroundColor": "#bbb",
              "borderRadius": 8,
              "height": 8,
              "width": 8,
            },
            Object {
              "marginHorizontal": 3,
              "marginVertical": 3,
            },
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 8,
              "width": 8,
            },
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "backgroundColor": "#ff0",
            },
          ]
        }
      />
      <View
        key="dot-1"
        style={
          Array [
            Object {
              "backgroundColor": "#bbb",
              "borderRadius": 8,
              "height": 8,
              "width": 8,
            },
            Object {
              "marginHorizontal": 3,
              "marginVertical": 3,
            },
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 8,
              "width": 8,
            },
            false,
            false,
          ]
        }
      />
      <View
        key="dot-2"
        style={
          Array [
            Object {
              "backgroundColor": "#bbb",
              "borderRadius": 8,
              "height": 8,
              "width": 8,
            },
            Object {
              "marginHorizontal": 3,
              "marginVertical": 3,
            },
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 8,
              "width": 8,
            },
            false,
            false,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Carousel Component renders correctly 2`] = `""`;

exports[`Carousel Component renders correctly with no child 1`] = `""`;

exports[`Carousel Component renders correctly with one child 1`] = `
<View
  onLayout={[Function]}
>
  <UnimplementedView
    accessibilityLabel="Carousel"
    autoplay={false}
    autoplayInterval={2000}
    bounces={true}
    carouselChange={
      [MockFunction] {
        "calls": Array [
          Array [
            2,
          ],
          Array [
            0,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    dotActiveStyle={
      Object {
        "backgroundColor": "#ff0",
      }
    }
    dotStyle={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 4,
        "height": 8,
        "width": 8,
      }
    }
    dots={[Function]}
    hasDots={true}
    initialPage={1}
    key="$carousel"
    loop={true}
    onPageScrollStateChanged={[Function]}
    onPageSelected={[Function]}
    pageStyle={Object {}}
    selectedIndex={0}
    style={
      Object {
        "height": 180,
      }
    }
    useViewPagerOnAndroid={true}
  >
    <View
      style={
        Array [
          Object {},
          Object {
            "width": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "red",
            "flex": 1,
          }
        }
      >
        <TYText
          align={null}
          color={null}
          size={null}
          style={
            Object {
              "color": "#333",
            }
          }
          type={null}
          weight={null}
        >
          Carousel 1
        </TYText>
      </View>
    </View>
  </UnimplementedView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "bottom": 10,
          "left": 0,
          "position": "absolute",
          "right": 0,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        key="dot-0"
        style={
          Array [
            Object {
              "backgroundColor": "#bbb",
              "borderRadius": 8,
              "height": 8,
              "width": 8,
            },
            Object {
              "marginHorizontal": 3,
              "marginVertical": 3,
            },
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 4,
              "height": 8,
              "width": 8,
            },
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "backgroundColor": "#ff0",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
